tests/functional/conftest.py

```diff
@@ -0,0 +1,268 @@
+
+            except Exception as e:
+                if conn.handle and conn.handle.closed == 0:
+                    conn.handle.rollback()
+                print(sql)
+                print(e)
+                raise
+            finally:
+                conn.transaction_open = False
+
+
+# Start at directory level.
+def write_project_files(project_root, dir_name, file_dict):
+    path = project_root.mkdir(dir_name)
+    if file_dict:
+        write_project_files_recursively(path, file_dict)
+
+
+# Write files out from file_dict. Can be nested directories...
+def write_project_files_recursively(path, file_dict):
+    for name, value in file_dict.items():
+        if name.endswith('.sql') or name.endswith('.csv') or name.endswith('.md'):
+            path.join(name).write(value)
+        elif name.endswith('.yml') or name.endswith('.yaml'):
+            if isinstance(value, str):
+                data = value
+            else:
+                data = yaml.safe_dump(value)
+            path.join(name).write(data)
+        else:
+            write_project_files_recursively(path.mkdir(name), value)
+
+
+@pytest.fixture
+def models():
+    return {}
+
+
+@pytest.fixture
+def macros():
+    return {}
+
+
+@pytest.fixture
+def seeds():
+    return {}
+
+
+@pytest.fixture
+def snapshots():
+    return {}
+
+
+@pytest.fixture
+def tests():
+    return {}
+
+
+@pytest.fixture
+def project_files(project_root, models, macros, snapshots, seeds, tests):
+    write_project_files(project_root, 'models', models)
+    write_project_files(project_root, 'macros', macros)
+    write_project_files(project_root, 'snapshots', snapshots)
+    write_project_files(project_root, 'seeds', seeds)
+    write_project_files(project_root, 'tests', tests)
+
+
+class TestProjInfo:
+    def __init__(
+        self, project_root, profiles_dir, adapter, test_dir, data_dir, test_schema, database
+    ):
+        self.project_root = project_root
+        self.profiles_dir = profiles_dir
+        self.adapter = adapter
+        self.test_dir = test_dir
+        self.data_dir = data_dir
+        self.test_schema = test_schema
+        self.database = database
+
+
+@pytest.fixture
+def project(
+    project_root,
+    profiles_root,
+    request,
+    unique_schema,
+    profiles_yml,
+    dbt_project_yml,
+    packages_yml,
+    selectors_yml,
+    schema,
+    project_files,
+    data_dir,
+):
+    # Return whatever is needed later in tests but can only come from fixtures, so we can keep the signatures
+    # in the test signature to a minimum. Also... database? ...?
+    return TestProjInfo(
+        project_root=project_root,
+        profiles_dir=profiles_root,
+        adapter=schema,
+        test_dir=request.fspath.dirname,
+        data_dir=data_dir,
+        test_schema=unique_schema,
+        # the following feels kind of fragile. TODO: better way of getting database
+        database=profiles_yml['test']['outputs']['default']['dbname'],
```

Comment:
This behavior is going to vary across adapters:
- Postgres cannot read from other databases
- Redshift cannot write to other databases (can sometimes read from others)
- BigQuery + Snowflake can read from + write to other databases/projects, it's important that we can test that functionality
- Spark/Databricks (currently) has no notion of a separate "database" level, it's just a synonym for `schema`

Reply:
Yeah, this is one of the ragged edges. Some other ragged edges are getting the adapter in a good way, and the quoting config.